File: identity_service/src/main/java/com/ticketflow/identity_service/service/UserService.java
package com.ticketflow.identity_service.service;

import java.util.Optional;

import com.ticketflow.identity_service.api.client_models.LoginModel;
import com.ticketflow.identity_service.domain.UserRepository;
import com.ticketflow.identity_service.models.User;
import com.ticketflow.identity_service.models.exceptions.NotFoundException;
import com.ticketflow.identity_service.models.exceptions.NotUniqueEntityException;
import com.ticketflow.identity_service.models.exceptions.WrongPasswordException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class UserService {
    private static final String WRONG_PASSWORD_EXCEPTION_MESSAGE = "Wrong password for user with email=%s";
    
    private UserRepository userRepository;
    private JwtGenerator jwtGenerator;

    @Autowired
    public UserService(UserRepository userRepository, JwtGenerator jwtGenerator) {
        this.userRepository = userRepository;
        this.jwtGenerator = jwtGenerator;
    }

    public User getByToken(String token) throws NotFoundException {
        return userRepository.getByToken(token);
    }

    public User getByEmail(String email) throws NotFoundException {
        return userRepository.getByEmail(email);
    }

    public String login(LoginModel loginModel) throws NotFoundException, WrongPasswordException {
        User user = userRepository.getByEmail(loginModel.getEmail());
        
        if (!user.getPassword().equals(loginModel.getPassword())) {
            throw new WrongPasswordException(String.format(WRONG_PASSWORD_EXCEPTION_MESSAGE, loginModel.getEmail()));
        }

        String newJwtToken = jwtGenerator.generate(user);
        user.setToken(Optional.of(newJwtToken));
        userRepository.update(user);
        return newJwtToken;
    }

    public void register(User user) throws NotUniqueEntityException {
        userRepository.add(user);
    }

    public void logout(String token) throws NotFoundException {
        User user = userRepository.getByToken(token);
        user.setToken(Optional.empty());
        userRepository.update(user);
    }
}

File: backend/dotnet/ticketflow/TicketFlow.IdentityService/Service/UserService.cs
using TicketFlow.IdentityService.Client.Entities;
using TicketFlow.IdentityService.Client.Extensibility.Entities;
using TicketFlow.IdentityService.Client.Extensibility.Exceptions;
using TicketFlow.IdentityService.Client.Extensibility.Factories;
using TicketFlow.IdentityService.Client.Extensibility.Models;
using TicketFlow.IdentityService.Persistence;

namespace TicketFlow.IdentityService.Service
{
    internal class UserService : IUserService
    {
        private readonly IUserRepository userRepository;
        private readonly IUserFactory userFactory;

        public UserService(IUserRepository userRepository, IUserFactory userFactory)
        {
            this.userRepository = userRepository;
            this.userFactory = userFactory;
        }

        public IAuthorizedUser GetByToken(string token)
        {
            if (userRepository.TryGetByToken(token, out IAuthorizedUser user))
            {
                return user;
            }

            throw new NotFoundException($"User with token={token} is not found");
        }

        public IUser GetByEmail(string email)
        {
            if (userRepository.TryGet(email, out IUser user))
            {
                return user;
            }

            throw new NotFoundException($"User with email={email} is not found");
        }

        public string Login(LoginRequest loginRequest)
        {
            IUser user = GetByEmail(loginRequest.Email);

            if (!user.TryAuthorize(loginRequest.Password, out IAuthorizedUser authorizedUser))
            {
                throw new WrongPasswordException($"Wrong password for user with email=${loginRequest.Email}");
            }

            userRepository.Update(authorizedUser);
            return authorizedUser.Token;
        }

        public void Register(RegisterRequest registerRequest)
        {
            if (userRepository.TryGet(registerRequest.Email, out _))
            {
                throw new NotUniqueEntityException($"User with email={registerRequest.Email} already exists");
            }

            IUser user = userFactory.Create(new UserCreationModel(registerRequest.Email, registerRequest.Password, Role.User));
            userRepository.Add(user);
        }

        public void Logout(string token)
        {
            IUser storedUser = GetByToken(token);
            IUser loggedOutUser = userFactory.Create(new UserCreationModel(storedUser.Email, storedUser.Password, storedUser.Role));

            userRepository.Update(loggedOutUser);
        }
    }
}